Phase 2: Documentation (#131)
Ekumen-OS/lambkin: README.md

# LAMBKIN

## Overview

LAMBKIN (Localization And Mapping Benchmarking) is a programmatic SDK for building reproducible, structured, and parallelized SLAM evaluation pipelines.
It moves away from complex automation "glue" in favor of a clean, Python-first approach to benchmarking.

## Scope

In its current phase, LAMBKIN acts as a sequential process orchestrator. It provides a streamlined, Python-first approach to the fundamental lifecycle management of ROS2-based benchmarking through strictly ordered execution, synchronized playback, and graceful termination of distributed nodes.

The logic handles node lifecycle management to ensure a clean ROS2 graph between iterations, synchronizes data flow between bag playback and node processing, and automates evaluation by interfacing with tools like evo to transform raw logs into standardized metrics.

While Lambkin is algorithm-agnostic by design and can integrate with any ROS2-based pipeline, this repository provides a specific worked example using the [Beluga](https://github.com/Ekumen-OS/beluga) AMCL localization algorithm, including a predefined launch file and configuration files to run a complete benchmark out of the box.

> ### ROS2 Beluga Example

Reviewer: ```suggestion
## ROS2 Beluga Example
```

Author: Solved in 714e8f0

# LAMBKIN

## Overview

LAMBKIN (Localization And Mapping Benchmarking) is a programmatic SDK for building reproducible, structured, and parallelized SLAM evaluation pipelines.
It moves away from complex automation "glue" in favor of a clean, Python-first approach to benchmarking.

## Scope

In its current phase, LAMBKIN acts as a sequential process orchestrator. It provides a streamlined, Python-first approach to the fundamental lifecycle management of ROS2-based benchmarking through strictly ordered execution, synchronized playback, and graceful termination of distributed nodes.

The logic handles node lifecycle management to ensure a clean ROS2 graph between iterations, synchronizes data flow between bag playback and node processing, and automates evaluation by interfacing with tools like evo to transform raw logs into standardized metrics.

> While Lambkin is algorithm-agnostic by design and can integrate with any ROS2-based pipeline, this repository provides a specific worked example using the [Beluga](https://github.com/Ekumen-OS/beluga) AMCL localization algorithm, including a predefined launch file and configuration files to run a complete benchmark out of the box.

Reviewer: Remove please as this has been moved to the section below.

Author: Solved in b2c0b4b

## ROS2 Beluga Example

> While Lambkin is algorithm-agnostic by design and can integrate with any ROS2-based localization algorithm, this repository provides a specific worked example using [Beluga](https://github.com/Ekumen-OS/beluga) AMCL. It includes a predefined ROS2 package with a launch file and configuration files ready to run a complete benchmark out of the box.

Reviewer: Some rephrase and let's mention docker.
```suggestion
While Lambkin is algorithm-agnostic by design and can integrate with any localization package, this repository provides a specific worked example using [Beluga](https://github.com/Ekumen-OS/beluga) AMCL. It includes a ROS2 package with a launch file and a configuration file ready to run a complete benchmark out of the box.

To simplify deployment, the repository provides a dockerized setup (see [docker](docker/) folder) that includes all necessary dependencies and tools pre-configured for immediate use.
```

Author: Solved in b2c0b4b

# LAMBKIN

## Overview

LAMBKIN (Localization And Mapping Benchmarking) is a programmatic SDK for
building reproducible, structured, and parallelized SLAM evaluation pipelines.
It moves away from complex automation "glue" in favor of a clean, Python-first

> approach to benchmarking.

Reviewer: Why do you split in several lines? Not just this paragrpah but the entire file.

Author: Solved in e5c21c

> ### What it does in this phase

Reviewer: ```suggestion
### Scope
```

Author: Solved in 33460a1

In this initial phase, LAMBKIN provides a simple Python script that serves as
the foundation for the orchestration system. It automates the full benchmarking
pipeline using **Beluga AMCL** as the localization algorithm,

> evaluating it across multiple configurations(sensor models and particle counts)

Reviewer: ```suggestion
evaluating it across multiple configurations (sensor models and particle counts)
```

Author: Solved in 33460a1

and measuring its accuracy against a ground truth trajectory.

- Launches Beluga AMCL with different configurations (sensor models, particle counts)
- Plays a reference rosbag to simulate a real robot environment
- Records the output topics and evaluates the estimated trajectory
- Computes the Absolute Pose Error (APE) against the ground truth

> - Generates comparison plots across all configurations

Reviewer: Rather than focusing on what the script does, which I believe is part of the architecture you explain below, I would focus on the scope of this phase. Do not be specific with Beluga (this is something we as users chose, but we could be using other algorithm).

Explain that it does a simple and sequential orchestration with some basic process lifecycle/termination  management. And that we provide a specific example with Beluga.

Author: Change some descriptions e5c21cf

## Architecture

### How it works

> Each benchmark iteration follows this sequence:

Reviewer: It is clear to see in this sequence where the loop for iteratons and variations happen. Might be worth doing a UML/flow diagram with boxes and arrows that depicts this.

Author: Added in 33460a1

```
1. Clean up any orphaned ROS2 nodes   (kill leftover processes from previous runs)
        ↓

> 2. Launch Beluga AMCL node            (ros2 launch → beluga_amcl + map_server + lifecycle_manager)

Reviewer: You not only launch beluga AMCL node. Better rephrase as Launch Beluga Algorithm as in braces you already mention the nodes that are launched).

Author: Solved in 33460a1

        ↓
4. Play reference rosbag              (ros2 bag play → publishes sensor data)
        ↓
5. Wait for playback to end           (synchronizes all process termination)
        ↓
6. Clean up all ROS2 nodes            (kill all processes before next iteration)
        ↓
7. Evaluate results                   (evo_traj → TUM format, evo_ape → APE metrics)
        ↓
8. Generate aggregated plots          (evo_res → APE comparison plots across all configurations)
```

### Use Case

LAMBKIN is designed to work with any localization algorithm package. In this
phase, it orchestrates the following ROS2 nodes:

- **Beluga AMCL** — the algorithm being benchmarked (e.g. a AMCL-based node),
  responsible for estimating the robot pose from sensor data and a known map
- **map_server** — provides the static map to the localization node
- **lifecycle_manager** — manages the lifecycle of both the localization node and
  map_server, handling their startup and shutdown transitions automatically

These three nodes are launched together via a ROS2 launch file, which accepts
parameters such as the map path, sensor model type, and maximum number of

> particles, allowing LAMBKIN to evaluate different configurations automatically.

Reviewer: LAMBKIN does not orchestrate nodes. It orchestrates processes, and particularly a launch process to run Beluga algorithm.

I like the comment that LAMBKIN can work with any algotirhm. But explain that we provide an example ros2 package with predefined launch and configuration files that allow to run Beluga. If you want, you can explain the dependencies of the ROS 2 package (the different nodes from different packages).

Author: Change the descriptions in e5c21cf

### Output folder structure

Lambkin automatically creates the following directory structure for all generated artifacts:

```
benchmarking_results/
└── <sensor_model>_p<num_particles>/
    └── iter_<number>/
        ├── bag/        ← Recorded rosbags
        ├── tum/        ← TUM trajectory files
        └── ape/        ← APE metrics and zip results
└── plots_ape/          ← Aggregated APE plots across all configurations
```

---

## Installation

### Prerequisites

Make sure you have the following installed:

- [Docker](https://docs.docker.com/get-docker/) and [Docker Compose](https://docs.docker.com/compose/)

### Expected artifacts

Make sure you have the following reference files available before running the benchmark:

| Artifact | Description |
|---|---|
| Rosbag | Reference sensor data to replay during the benchmark |

> | Map | Static map file in `.yaml` format |

Reviewer: Not only yaml but the associated pgm.

Author: Solved in 33460a1

> | Map | Static map file in `.yaml` format |
> | Groundtruth | Reference trajectory in `.tum` format to evaluate against |
> ### Setup

Reviewer: ```suggestion
| Groundtruth | Reference trajectory in `.tum` format to evaluate against |

### Setup
```

Author: Solved in e5c21cf

---

## Usage

### Running the benchmark

```bash
uv run src/lambkin/lambkin.py
```

### Dry mode

To preview all commands without executing them, set `DRY_MODE = True` in `src/lambkin/lambkin.py`:

```python
DRY_MODE = True
```

Then run normally — it will print all commands to stdout without launching any processes.

### Interpreting results

After the benchmark completes:

- **APE metrics** are saved as `.zip` files under each `iter_<n>/ape/` directory
- **TUM trajectories** are saved under each `iter_<n>/tum/` directory
- **Aggregated APE plots** comparing all configurations are saved under `benchmarking_results/plots_ape/`

The APE (Absolute Pose Error) metric measures the difference between the estimated trajectory and the ground truth. Lower APE values indicate better localization accuracy.

## Example Results

Plot generated after running the benchmark with all configurations.

![APE Comparison Plot](doc/plot_ape.png)

---

## Known Bags

- Improve process management to reliably terminate all ROS2 child processes across iterations
- Fix orphaned processes that survive after a crashed run and another iteration

> - Fix tight_layout warning in evo_res plots and overlapping axis labels

Reviewer: Add actual message:
```bash
  /ws/.venv/lib/python3.12/site-packages/seaborn/axisgrid.py:123: UserWarning: The figure layout has changed to tight
  self._figure.tight_layout(*args, **kwargs)
```

Author: Added in a81a99e

# LAMBKIN

## Overview

LAMBKIN (Localization And Mapping Benchmarking) is a programmatic SDK for building reproducible, structured, and parallelized SLAM evaluation pipelines.
It moves away from complex automation "glue" in favor of a clean, Python-first approach to benchmarking.

### Scope

> Lambkin is a simple, sequential process orchestrator that covers the basic concepts of process lifecycle management and graceful termination for ROS2-based benchmarking pipelines. It handles each stage of a benchmark run in a strictly ordered sequence — spawning nodes, recording topics, synchronizing playback, and collecting evaluation metrics — making the execution flow explicit and easy to follow.

Reviewer: ```suggestion
In its current phase, LAMBKIN acts as a sequential process orchestrator. It provides a streamlined, Python-first approach to the fundamental lifecycle management of ROS2-based benchmarking through strictly ordered execution, synchronized playback, and graceful termination of distributed nodes. 

The logic handles node lifecycle management to ensure a clean ROS2 graph between iterations, synchronizes data flow between bag playback and node processing, and automates evaluation by interfacing with tools like evo to transform raw logs into standardized metrics.
```

Author: Solved in 6de27e6

> While Lambkin is algorithm-agnostic by design and can integrate with any ROS2-based pipeline, this repository provides a specific worked example using the [Beluga](https://github.com/Ekumen-OS/beluga) AMCL localization algorithm, including a predefined launch file and configuration files to run a complete benchmark out of the box.

Reviewer: Would move this and `### ROS2 Package` section to a unified section in level 2 headers (##). You can name it ROS 2 Beluga Example. Can be on top of ## Architecture.

Author: Solved in 6de27e6

Author: Solved name section 4ca5f8f

Reviewer: This is level 3 header, not 2. Please fix

    D["<b>4. Play Reference Rosbag</b><br>ros2 bag play reference.bag<br>Publishes sensor data into live ROS2 graph"]
    E["<b>5. Wait for Playback to End</b><br>wait $BAG_PID<br>Synchronizes all process termination"]
    F["<b>6. Clean up All ROS2 Nodes</b><br>kill $LAUNCH_PID $RECORD_PID<br>Kill all processes before next iteration"]
    G["<b>7. Evaluate Results</b><br>evo_traj tum ... · evo_ape tum ...<br>TUM format conversion · APE metrics"]
    H["<b>8. Generate Aggregated Plots</b><br>evo_res results/*.zip --plot<br>APE comparison plots across all configurations"]

    END(( ))

    A --> B --> C --> D --> E --> F --> G --> H

    style START fill:#111827,stroke:#111827
    style END   fill:#111827,stroke:#111827

    style A fill:#fee2e2,stroke:#dc2626,color:#7f1d1d
    style B fill:#dbeafe,stroke:#2563eb,color:#1e3a8a
    style C fill:#dcfce7,stroke:#16a34a,color:#14532d
    style D fill:#fef3c7,stroke:#d97706,color:#78350f
    style E fill:#f3f4f6,stroke:#6b7280,color:#1f2937
    style F fill:#fee2e2,stroke:#dc2626,color:#7f1d1d
    style G fill:#ede9fe,stroke:#7c3aed,color:#4c1d95
    style H fill:#ede9fe,stroke:#7c3aed,color:#4c1d95

> ```

Reviewer: I am not convinced on how this is rendered. Boxes and style is too big. Also, it does not reflect how the iterations and variations happen. You can use pseudocode if needed. Consider using draw.io as well to draw the diagrams as it might be easier.

Author: I create new diagram in 6de27e6

| `PROCESS_TERMINATION_TIMEOUT` | `7.0` | Seconds before force-killing a process |

The benchmark will automatically run **all combinations** of `LASER_MODELS` × `NUM_PARTICLES` × `NUM_ITERATIONS`.

---

## Usage

### Running the benchmark

```bash
uv run src/lambkin/lambkin.py
```

### Dry mode

To preview all commands without executing them, set `DRY_MODE = True` in `src/lambkin/lambkin.py`:

```python
DRY_MODE = True
```

Then run normally — it will print all commands to stdout without launching any processes.

### Interpreting results

After the benchmark completes:

- **APE metrics** are saved as `.zip` files under each `iter_<n>/ape/` directory
- **TUM trajectories** are saved under each `iter_<n>/tum/` directory
- **Aggregated APE plots** comparing all configurations are saved under `benchmarking_results/plots_ape/`

The APE (Absolute Pose Error) metric measures the difference between the estimated trajectory and the ground truth. Lower APE values indicate better localization accuracy.

## Example Results

Plot generated after running the benchmark with all configurations.

![APE Comparison Plot](doc/plot_ape.png)

---

> ## Known Bags

Reviewer: ```suggestion
## Known Bugs
```

Author: Solved in 6de27e6

- Improve process management to reliably terminate all ROS2 child processes across iterations
- Fix orphaned processes that survive after a crashed run and another iteration
- Fix tight_layout warning in evo_res plots and overlapping axis labels

        ```
        /ws/.venv/lib/python3.12/site-packages/seaborn/axisgrid.py:123: UserWarning: The figure layout has changed to tight self._figure.tight_layout
        (*args **kwargs)

>         ```

Reviewer: Indentation is not correct. See how it renders. Also, fenced code blocks (```) should have a language specified

<img width="1043" height="196" alt="Image" src="https://github.com/user-attachments/assets/3634e430-91eb-4d67-b031-1e55c0c2f8c1" />

Author: Solved in 6de27e6
